### ISSUE TITLE
Cannot start the zookeeper client in a Node.js Worker thread

### ISSUE THREAD

**Issue by tszyszko** [yfinkelstein/node-zookeeper#281]
**Describe the bug**

Running within a worker thread inside docker causes node-zookeeper to crash on server connect:

**To Reproduce**
Steps to reproduce the behaviour:

Run zookeeper in a worker thread.

file: test.mjs

```
import Zookeeper from 'zookeeper';

let zkClient;
const REQUEST_TIMEOUT = 5000;

function createClient() {
    if (!zkClient) {
      let zkClientConnected = false;
      console.info('creating a ZK Client');

      let zkClientConfig = {
        connect: 'localhost:2181',
        timeout: REQUEST_TIMEOUT,
        host_order_deterministic: false,
        debug_level: Zookeeper.ZOO_LOG_LEVEL_DEBUG,
      };

      let zkClientReadyPromise = new Promise(resolve => {
        zkClient = new Zookeeper(zkClientConfig);

        zkClient.on('close', () => {
          zkClientConnected = false;
          console.info('close', `session closed, id=${zkClient.client_id}`);
          zkClient = null;
        });

        zkClient.on('connecting', () => {
          zkClientConnected = false;
          console.info('connecting', `session connecting, id=${zkClient.client_id}`);
        });

        zkClient.on('connect', () => {
          zkClientConnected = true;
          console.info('connect', `session connect, id=${zkClient.client_id}`);
          resolve(true);
        });

        setTimeout(() => {
          zkClient.init(zkClientConfig);
        }, 1000);

      });

    }
  }

createClient();
```


file main.mjs:

```
import { MessageChannel, Worker } from 'worker_threads';
console.log('test');
function startWorker(path) {
  const w = new Worker(path, { workerData: null });
  w.on('message', msg => {
    logger.info(`worker : ${msg}`);
  });
  w.on('error', err => {
    logger.error(`worker : ${err}`);
  });
  w.on('exit', code => {
    if (code !== 0) {
      logger.error(`Worker stopped with exit code ${code}`);
    }
    // Wait a second for log messages to flush
    setTimeout(() => {
      process.exit(1);
    }, 1000);
  });
  return w;
}

const test = startWorker(new URL('test.mjs', import.meta.url));


await new Promise(resolve => setTimeout(resolve, 30000));
```

file Dockerfile:
```
FROM node:16-slim

WORKDIR /app

COPY . ./

RUN yarn

CMD ["node", "main.mjs" ]

```



**Expected behavior**
```
docker run --net host  -it zookeeper_test node test.mjs                                                      
creating a ZK Client                                                                                                                        
2021-10-15 16:25:14,328:1:ZOO_INFO@log_env@1102: Client environment:zookeeper.version=zookeeper C client 3.5.8                                                                         
2021-10-15 16:25:14,328:1:ZOO_INFO@log_env@1113: Client environment:os.name=Linux                                                            
2021-10-15 16:25:14,328:1:ZOO_INFO@log_env@1114: Client environment:os.arch=5.4.0-86-generic                                                 
2021-10-15 16:25:14,328:1:ZOO_INFO@log_env@1115: Client environment:os.version=#97-Ubuntu SMP Fri Sep 17 19:19:40 UTC 2021                   
2021-10-15 16:25:14,328:1:ZOO_INFO@log_env@1123: Client environment:user.name=(null)                                                         
2021-10-15 16:25:14,329:1:ZOO_INFO@log_env@1131: Client environment:user.home=/root                                                          
2021-10-15 16:25:14,329:1:ZOO_INFO@log_env@1143: Client environment:user.dir=/app                                                            
2021-10-15 16:25:14,329:1:ZOO_INFO@zookeeper_init_internal@1177: Initiating client connection, host=localhost:2181 sessionTimeout=5000 watcher=0x7fb8001fa4b0 sessionId=0 sessionPasswd=<null> context=0x5223be0 flags=0
2021-10-15 16:25:14,330:1:ZOO_DEBUG@get_next_server_in_reconfig@1330: [OLD] count=0 capacity=0 next=0 hasnext=0                              
2021-10-15 16:25:14,330:1:ZOO_DEBUG@get_next_server_in_reconfig@1333: [NEW] count=1 capacity=16 next=0 hasnext=1                             
2021-10-15 16:25:14,330:1:ZOO_DEBUG@get_next_server_in_reconfig@1344: Using next from NEW=127.0.0.1:2181                                     
2021-10-15 16:25:14,331:1:ZOO_DEBUG@zookeeper_connect@2232: [zk] connect()                                                                   

Interest in (fd=19, read=true, write=true, timeout=1666)yield: creating a new poll handle for 0x5223be0yield: starting poll for 0x5223be0 from thread 0x13zk_io_cb fired, status: 0, revents: 22021-10-15 16:25:14,332:1:ZOO_INFO@check_events@2473: initiated connection to server [127.0
.0.1:2181]                                                            
Interest in (fd=19, read=true, write=false, timeout=1666)yield: starting poll for 0x5223be0 from thread 0x7fb800bf28c0zk_io_cb fired, status: 0, revents: 12021-10-15 16:25:14,341:1:ZOO_INFO@check_events@2521: session establishment complete on server [127.0.0.1:2181], sessionId=0x10
06b8f657a0006, negotiated timeout=6000                                
2021-10-15 16:25:14,341:1:ZOO_DEBUG@check_events@2531: Calling a watcher for a ZOO_SESSION_EVENT and the state=ZOO_CONNECTED_STATE           
2021-10-15 16:25:14,342:1:ZOO_DEBUG@process_completions@2815: Calling a watcher for node [], type = -1 event=ZOO_SESSION_EVENT               
main watcher event: type=-1, state=3, path=calling Emit(connect, path='')connect session connect, id=1006b8f657a0006                         
Interest in (fd=19, read=true, write=false, timeout=1987)yield: starting poll for 0x5223be0 from thread 0x7fb800bf28c0zk_timer_cb firedping timer went offInterest in (fd=19, read=true, write=false, timeout=1)yield: starting poll for 0x5223be0 from thread 0x7fb800bf28c0zk_timer_cb f
ireddelaying ping timer by 1zk_timer_cb firedping timer went offInterest in (fd=19, read=true, write=false, timeout=2000)yield: starting poll for 0x5223be0 from thread 0x7fb800bf28c0zk_io_cb fired, status: 0, revents: 12021-10-15 16:25:16,334:1:ZOO_DEBUG@zookeeper_process@2898: Got
 ping response in 421052648 ms                                        
Interest in (fd=19, read=true, write=false, timeout=1999)yield: starting poll for 0x5223be0 from thread 0x7fb800bf28c0^Czk_timer_cb firedping timer went offInterest in (fd=19, read=true, write=false, timeout=2000)yield: starting poll for 0x5223be0 from thread 0x7fb800bf28c0zk_io_cb
 fired, status: 0, revents: 12021-10-15 16:25:18,336:1:ZOO_DEBUG@zookeeper_process@2898: Got ping response in 421052648 ms                   
Interest in (fd=19, read=true, write=false, timeout=2000)yield: starting poll for 0x5223be0 from thread 0x7fb800bf28c0^C   
```            

**Error Log**
```
docker run --net host  -it zookeeper_test node main.mjs                         
test                                                                                                                                        
creating a ZK Client                                                                                                                        
2021-10-15 16:25:07,632:1:ZOO_INFO@log_env@1102: Client environment:zookeeper.version=zookeeper C client 3.5.8                          
2021-10-15 16:25:07,632:1:ZOO_INFO@log_env@1113: Client environment:os.name=Linux                                                           
2021-10-15 16:25:07,632:1:ZOO_INFO@log_env@1114: Client environment:os.arch=5.4.0-86-generic                                                
2021-10-15 16:25:07,632:1:ZOO_INFO@log_env@1115: Client environment:os.version=#97-Ubuntu SMP Fri Sep 17 19:19:40 UTC 2021
2021-10-15 16:25:07,632:1:ZOO_INFO@log_env@1123: Client environment:user.name=(null)                                                        
2021-10-15 16:25:07,633:1:ZOO_INFO@log_env@1131: Client environment:user.home=/root                                                         
2021-10-15 16:25:07,633:1:ZOO_INFO@log_env@1143: Client environment:user.dir=/app                              
2021-10-15 16:25:07,633:1:ZOO_INFO@zookeeper_init_internal@1177: Initiating client connection, host=localhost:2181 sessionTimeout=5000 watcher=0x7fbd5011d4b0 sessionId=0 sessionPasswd=<null> context=0x7fbd3c0b5630 flags=0
2021-10-15 16:25:07,634:1:ZOO_DEBUG@get_next_server_in_reconfig@1330: [OLD] count=0 capacity=0 next=0 hasnext=0               
2021-10-15 16:25:07,634:1:ZOO_DEBUG@get_next_server_in_reconfig@1333: [NEW] count=1 capacity=16 next=0 hasnext=1    
2021-10-15 16:25:07,634:1:ZOO_DEBUG@get_next_server_in_reconfig@1344: Using next from NEW=127.0.0.1:2181                                    
2021-10-15 16:25:07,635:1:ZOO_DEBUG@zookeeper_connect@2232: [zk] connect()                                                                  
                                                                                                                                            
Interest in (fd=24, read=true, write=true, timeout=1666)yield: creating a new poll handle for 0x7fbd3c0b5630yield: starting poll for 0x7fbd3c0b5630 from thread 0x18

--- Segfaults --
```

When running puppeteer in the main thread, was able to get the following stack trace:
```
[info] -  [undefined] -  creating a ZK Client 
2021-10-15 15:02:27,020:10:ZOO_INFO@log_env@1102: Client environment:zookeeper.version=zookeeper C client 3.5.8
2021-10-15 15:02:27,021:10:ZOO_INFO@log_env@1106: Client environment:host.name=c4657b7f2050
2021-10-15 15:02:27,021:10:ZOO_INFO@log_env@1113: Client environment:os.name=Linux
2021-10-15 15:02:27,021:10:ZOO_INFO@log_env@1114: Client environment:os.arch=5.4.0-86-generic
2021-10-15 15:02:27,021:10:ZOO_INFO@log_env@1115: Client environment:os.version=#97-Ubuntu SMP Fri Sep 17 19:19:40 UTC 2021
2021-10-15 15:02:27,021:10:ZOO_INFO@log_env@1123: Client environment:user.name=(null)
2021-10-15 15:02:27,021:10:ZOO_INFO@log_env@1131: Client environment:user.home=/home/pptruser
2021-10-15 15:02:27,021:10:ZOO_INFO@log_env@1143: Client environment:user.dir=/app
2021-10-15 15:02:27,021:10:ZOO_INFO@zookeeper_init_internal@1186: Initiating client connection, host=zookeeper:2181 sessionTimeout=30000 watcher=0x7fa1a2f12150 sessionId=0 sessionPasswd=<null> context=0x7f9f0e6ec680 flags=0
2021-10-15 15:02:27,022:10:ZOO_DEBUG@get_next_server_in_reconfig@1332: [OLD] count=0 capacity=0 next=0 hasnext=0
2021-10-15 15:02:27,022:10:ZOO_DEBUG@get_next_server_in_reconfig@1335: [NEW] count=1 capacity=16 next=0 hasnext=1
2021-10-15 15:02:27,022:10:ZOO_DEBUG@get_next_server_in_reconfig@1344: Using next from NEW=10.0.0.30:2181
2021-10-15 15:02:27,022:10:ZOO_DEBUG@zookeeper_connect@2232: [zk] connect()

2021-10-15 15:02:32,488:10:ZOO_INFO@check_events@2473: initiated connection to server [10.0.0.30:2181]
2021-10-15 15:02:32,497:10:ZOO_INFO@check_events@2525: session establishment complete on server [10.0.0.30:2181], sessionId=0x1005bfe28fa0027, negotiated timeout=30000 
2021-10-15 15:02:32,497:10:ZOO_DEBUG@check_events@2531: Calling a watcher for a ZOO_SESSION_EVENT and the state=ZOO_CONNECTED_STATE
2021-10-15 15:02:32,497:10:ZOO_DEBUG@process_completions@2817: Calling a watcher for node [], type = -1 event=ZOO_SESSION_EVENT
FATAL ERROR: v8::HandleScope::CreateHandle() Cannot create a handle without a HandleScope
 1: 0xb03020 node::Abort() [node]
 2: 0xa181fb node::FatalError(char const*, char const*) [node]
 3: 0xced61a v8::Utils::ReportApiFailure(char const*, char const*) [node]
 4: 0xe58a02 v8::internal::HandleScope::Extend(v8::internal::Isolate*) [node]
 5: 0x10771e4 v8::internal::JSReceiver::GetCreationContext() [node]
 6: 0xcf7808 v8::Object::GetCreationContext() [node]
 7: 0xa3fa2a node::MakeCallback(v8::Isolate*, v8::Local<v8::Object>, char const*, int, v8::Local<v8::Value>*, node::async_context) [node]
 8: 0x7fa1a2f0ffa6 zk::ZooKeeper::DoEmitPath(v8::Local<v8::String>, char const*) [/app/node_modules/zookeeper/build/Release/zookeeper.node]
 9: 0x7fa1a2f1225a zk::ZooKeeper::main_watcher(_zhandle*, int, int, char const*, void*) [/app/node_modules/zookeeper/build/Release/zookeeper.node]
10: 0x7fa1a2f2f973  [/app/node_modules/zookeeper/build/Release/zookeeper.node]
11: 0x7fa1a2f2fb58 deliverWatchers [/app/node_modules/zookeeper/build/Release/zookeeper.node]
12: 0x7fa1a2f28763 process_completions [/app/node_modules/zookeeper/build/Release/zookeeper.node]
13: 0x7fa1a2f27a2d  [/app/node_modules/zookeeper/build/Release/zookeeper.node]
14: 0x7fa1a2f27652  [/app/node_modules/zookeeper/build/Release/zookeeper.node]
15: 0x7fa1a2f289a7 zookeeper_process [/app/node_modules/zookeeper/build/Release/zookeeper.node]
16: 0x7fa1a2f1054b zk::ZooKeeper::zk_io_cb(uv_poll_s*, int, int) [/app/node_modules/zookeeper/build/Release/zookeeper.node]
17: 0x156c854  [node]
18: 0x155ad18 uv_run [node]
19: 0xa3ffe5 node::SpinEventLoop(node::Environment*) [node]
20: 0xb451b6 node::NodeMainInstance::Run(node::EnvSerializeInfo const*) [node]
21: 0xac6952 node::Start(int, char**) [node]
22: 0x7fa1a2fe109b __libc_start_main [/lib/x86_64-linux-gnu/libc.so.6]
23: 0xa3bfac  [node]
Interest in (fd=26, read=true, write=true, timeout=10000)yield: creating a new poll handle for 0x7f9f0e6ec680yield: starting poll for 0x7f9f0e6ec680 from thread 0x1azk_io_cb fired, status: 0, revents: 2Interest in (fd=26, read=true, write=false, timeout=10000)yield: starting poll for 0x7f9f0e6ec680 from thread (nil)zk_io_cb fired, status: 0, revents: 1main watcher event: type=-1, state=3, path=calling Emit(connect, path='')

```



**Desktop (please complete the following information):**
 - OS: Docker image based on debian breezy
 - Node.js version: Node 16 (but happens in 14 too)
 - C/C++ compiler and version:

**Additional context**

**Comment by DavidVujic** [comment]
Apologies for not responding to this issue until now. I will investigate this one, thank you for reporting!

**Comment by DavidVujic** [comment]
Note to self: maybe a possible solution and/or ideas here: https://github.com/serialport/node-serialport/issues/1981

**Comment by DavidVujic** [comment]
Hello @tszyszko, 
I'm not sure how to enable Node.js worker thread support. I think it might be a bit more difficult than the examples I've found (switching out NODE_MODULE for NODE_MODULE_INIT in the addon wrapper).

Can you explain the use case a little bit more?

**Comment by DavidVujic** [comment]
@tszyszko This is now solved by @deepcoldy in #310 

Closing this issue.
